processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -603,63 +603,84 @@ private ConstructorAccessor getConstructorAccessor(Type type) {
             List<ExecutableElement> constructors = ElementFilter.constructorsIn( type.getTypeElement()
                 .getEnclosedElements() );
 
-            ExecutableElement defaultConstructor = null;
+            // The rules for picking a constructor are the following:
+            // 1. Constructor annotated with @Default (from any package) has highest precedence
+            // 2. If there is a single public constructor then it would be used to construct the object
+            // 3. If a parameterless constructor exists then it would be used to construct the object, and the other
+            // constructors will be ignored
+            ExecutableElement defaultAnnotatedConstructor = null;
+            ExecutableElement parameterLessConstructor = null;
             List<ExecutableElement> accessibleConstructors = new ArrayList<>( constructors.size() );
+            List<ExecutableElement> publicConstructors = new ArrayList<>( );
 
             for ( ExecutableElement constructor : constructors ) {
                 if ( constructor.getModifiers().contains( Modifier.PRIVATE ) ) {
                     continue;
                 }
 
+                if ( hasDefaultAnnotationFromAnyPackage( constructor ) ) {
+                    // We found a constructor annotated with @Default everything else is irrelevant
+                    defaultAnnotatedConstructor = constructor;
+                    break;
+                }
+
                 if ( constructor.getParameters().isEmpty() ) {
-                    defaultConstructor = constructor;
+                    parameterLessConstructor = constructor;
                 }
                 else {
                     accessibleConstructors.add( constructor );
                 }
+
+                if ( constructor.getModifiers().contains( Modifier.PUBLIC ) ) {
```

Comment:
Should we also allow friendly and protected?  I guess we have some logic somewhere to detect these scenarios...

Reply:
We do allow protected ones. Those are part of the `accessibleConstructors`. This is to have public constructors with e precedence, so we can check them first.